src/openclsim/critical_path/dependencies_from_simpy_step.py

```diff
@@ -1,40 +1,245 @@
+            From self.get_recorded_activity_df().
+        """
+        activity_id = self.step_logging_dataframe.loc[e_id, "event_object"].value
+        end_time = round(self.step_logging_dataframe.loc[e_id, "t1"], 4)
+        matching_ids = recorded_activities_df.loc[
+            (
+                (recorded_activities_df.ActivityID == activity_id)
+                & (recorded_activities_df.end_time == end_time)
+            ),
+            "cp_activity_id",
+        ]
+        if len(set(matching_ids)) == 1:
+            cp_activity_id = matching_ids.iloc[0]
+        else:
+            raise ValueError(
+                f"No recorded activity found for"
+                f" activityID {activity_id} at (end)time {end_time}"
+            )
+        return cp_activity_id
+
+
+class AlteredStepEnv(simpy.Environment):
+    """
+    Class is child of simpy.Environment and passes on all arguments on initialization.
+    The 'step' method is overwritten (or 'monkey-patched') in order to log some data of
+    simulation into self.data_step and self.data_cause_effect. The former saves some metadata
+    of the Event such as e_id (execution ID), simulation time, prio and event type (list of tuples).
+    The latter saves which e_id scheduled another e_id and is hence a list of cause-effect tuples.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialization."""
+        super().__init__(*args, **kwargs)
+        self.data_cause_effect = []
+        self.data_step = []
+
+    def step(self):
+        """
+        The 'step' method is overwritten (or 'monkey-patched') in order to log some data of
+        simulation into self.data_step and self.data_cause_effect.
+        """
+        time_start = copy.deepcopy(self.now)
+        if len(self._queue):
+            _, prio, e_id, event = self._queue[0]
+            old_e_ids = set([t[2] for t in self._queue])
+        else:
+            _, prio, e_id, event = None, None, None, None
+            old_e_ids = {}
+
+        super().step()
+
+        if len(self._queue):
+            new_e_ids = list(set([t[2] for t in self._queue]) - old_e_ids)
+        else:
+            new_e_ids = []
+
+        time_end = copy.deepcopy(self.now)
+
+        self._monitor_cause_effect(e_id, new_e_ids)
+        self._monitor_step(time_start, time_end, prio, e_id, event)
+
+    def _monitor_cause_effect(self, e_id_current, e_ids_new):
+        """
+        Append dependencies (triggers) to data_cause_effect.
+
+        Parameters
+        ----------
+        e_id_current : int
+            Simpy execution ID (cause).
+        e_ids_new : list
+            Simpy execution IDs (effect).
+            If empty list, eid_current does not trigger another event.
+        """
+        if len(e_ids_new) > 0:
+            for new_eid in e_ids_new:
+                self.data_cause_effect.append((e_id_current, new_eid))
+
+    def _monitor_step(self, t0, t1, prio, e_id, event):
```

Comment:
maybe ``t_before`` and ``t_after``, or ``t_start`` and ``t_end``?

Reply:
yes